### PR TITLE
feat(server): Implement BITPOS.

### LINKED ISSUE #835: BITPOS does not work
Currently, `BITPOS` has no implementation, in the sense that the command is set to always return zero. 

https://github.com/dragonflydb/dragonfly/blob/f1751278378a9b067d0b744512d68c703f99ac38/src/server/bitops_family.cc#L485-L487

**Example**
```shell
> SET mykey "\xff\x00"
> BITPOS mykey 0
0
```
whereas it is expected to return 8.

### PR DESCRIPTION
**Pull request by chakaz**
Fixes #835.

### CODE REVIEW

File: src/server/bitops_family.cc [hunk 1]
```diff
@@ -685,6 +748,90 @@ OpResult<std::size_t> CountBitsForValue(const OpArgs& op_args, std::string_view
...
+    position =
+        FindFirstBitWithValueAsBit(result_value, bit_value, normalized_start, normalized_end);
+  } else {
+    position =
+        FindFirstBitWithValueAsByte(result_value, bit_value, normalized_start, normalized_end);
+  }
+
+  if (position == -1 && !bit_value && static_cast<size_t>(start) < result_value.size() &&
+      end == std::numeric_limits<int64_t>::max()) {
+    // Not-found return value is compatible with Redis, but is subtle.
```
Comment (ashotland) [inline src/server/bitops_family.cc:828]: The comment is a bit enigmatic to me :) 

If returning -1 is compatible with Redis I am not sure I agree with the logic, or with the cognitive burden on the reader this edge-case yields.

File: src/server/bitops_family.cc [hunk 2]
```diff
@@ -685,6 +748,90 @@ OpResult<std::size_t> CountBitsForValue(const OpArgs& op_args, std::string_view
...
+
+    return i * OFFSET_FACTOR + GetFirstBitWithValueInByte(current_byte, bit_value);
+  }
+
+  return -1;
+}
+
+OpResult<int64_t> FindFirstBitWithValue(const OpArgs& op_args, std::string_view key, bool bit_value,
+                                        int64_t start, int64_t end, bool as_bit) {
+  OpResult<std::string> result = ReadValue(op_args.db_cntx, key, op_args.shard);
```
Comment (ashotland) [inline src/server/bitops_family.cc:799]: rename result to value?

Result is a bit confusing with the result of the function.

File: src/server/bitops_family.cc [hunk 3]
```diff
@@ -685,6 +748,90 @@ OpResult<std::size_t> CountBitsForValue(const OpArgs& op_args, std::string_view
...
+
+  std::string_view result_value;
```
Comment (ashotland) [inline src/server/bitops_family.cc:801]: Same here, could be input_value or similar.

File: src/server/bitops_family_test.cc [hunk 4]
```diff
@@ -420,4 +420,115 @@ TEST_F(BitOpsFamilyTest, BitOpsNot) {
...
+  EXPECT_EQ(21, CheckedInt({"bitpos", "a", "1", "0", "21", "BIT"}));
+  EXPECT_EQ(21, CheckedInt({"bitpos", "a", "1", "21", "21", "BIT"}));
+  EXPECT_EQ(21, CheckedInt({"bitpos", "a", "1", "21", "100", "BIT"}));
+  EXPECT_EQ(21, CheckedInt({"bitpos", "a", "1", "0", "100", "BIT"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "a", "1", "-1", "-1", "BIT"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "a", "1", "-4", "-1", "BIT"}));
+  EXPECT_EQ(35, CheckedInt({"bitpos", "a", "1", "-5", "-1", "BIT"}));
+  EXPECT_EQ(34, CheckedInt({"bitpos", "a", "1", "-6", "-1", "BIT"}));
+
+  // Test weird return value spec, specifically when looking for clear bits in
```
Comment (ashotland) [inline src/server/bitops_family_test.cc:513]: // Edge cases for for clear bits in an all-set string.

**Review by ashotland — COMMENTED** [review]
Thanks! - just some naming related comments

File: src/server/bitops_family.cc (hunk 1 above)
Comment (chakaz) [reply src/server/bitops_family.cc:828]: In this specific case, Redis returns (and documents that it returns) the size, in bits, of the value.
I agree, this edge case is super weird and has a potentially significant burden on users. But that's the Redis API, and Dragonfly aims to be compatible, so we must mimic this weird API :(
I slightly modified the comment, please lmk if that's not good enough or if you have other suggestions.

File: src/server/bitops_family.cc (hunk 2 above)
Comment (chakaz) [reply src/server/bitops_family.cc:799]: I wanted to distinguish `bit_value` from `value`, but you're right that it's more confusing as is. Renamed to `value` per your suggestion.

File: src/server/bitops_family.cc (hunk 3 above)
Comment (chakaz) [reply src/server/bitops_family.cc:801]: I went with `value_str` to reflect that it's the value's string value. lmk if you have a better idea.

File: src/server/bitops_family_test.cc (hunk 4 above)
Comment (chakaz) [reply src/server/bitops_family_test.cc:513]: It's actually not the edge case that I'm testing, but rather that we behave like Redis in terms of the weird spec.
I updated the comment.

**Review by ashotland — DISMISSED** [review]
Thanks!

File: src/server/bitops_family_test.cc [hunk 5]
```diff
@@ -420,4 +420,114 @@ TEST_F(BitOpsFamilyTest, BitOpsNot) {
...
+  EXPECT_EQ(24, CheckedInt({"bitpos", "b", "0", "0"}));
+  EXPECT_EQ(24, CheckedInt({"bitpos", "b", "0", "1"}));
+  EXPECT_EQ(24, CheckedInt({"bitpos", "b", "0", "2"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "b", "0", "3"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "b", "0", "0", "1"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "b", "0", "0", "1", "BYTE"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "b", "0", "0", "3"}));
+  EXPECT_EQ(-1, CheckedInt({"bitpos", "b", "0", "0", "3", "BYTE"}));
+
+  ASSERT_EQ(Run({"set", "c", ""_b}), "OK");
```
[the remark this answers is not in the document]
Comment (chakaz) [reply src/server/bitops_family_test.cc]: Done :)

**Review by ashotland — APPROVED** [review]

**Pull request merged**